### PR TITLE
Basic Implementation for Towers

### PR DESCRIPTION
**Pull request by darshan3**
**Project Cards**:
1. [Towers [1]](https://github.com/orgs/Terasology/projects/17#card-21938530)

**How to test**:  
1. Start a new game with Core or Builder Gameplay mode.
2. Give yourself a Tower Root block through command `give towerRoot`
3. Place it and interact with it.
4. Activate the tower

**Expected outcome**:
A screen should appear with an activation checkbox. Click it to activate the towers. A prefab should appear above the root block to indicate that it is activated.

### CODE REVIEW

File: src/main/java/org/terasology/logic/ingame/towers/TowerComponent.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.terasology.logic.ingame.towers;
+
+import org.terasology.entitySystem.Component;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.network.Replicate;
+import org.terasology.world.block.ForceBlockActive;
+
+@ForceBlockActive
+public class TowerComponent implements Component {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/ingame/towers/TowerComponent.java]: Again, JavaDoc is highly appreciated (but can be added in a later PR).

**Review by skaldarnar — CHANGES_REQUESTED** [review]
Please look into the `testRook.prefab` and whether it is really needed. Otherwise the code looks good, documentation could also be added in a follow-up PR.

File: assets/blocks/towerRoot.block [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "displayName" : "Tower Root",
+    "hardness" : 10,
+    "tile" : "Core:HardStone",
+    "entity" : {
+        "prefab" : "Towers:towerRoot"
+    },
+    "attachmentAllowed": false,
```
[the remark this answers is not in the document]
Comment (darshan3) [reply assets/blocks/towerRoot.block]: Yes, towers in GooeyDefence support multi-block variants, that is, you can place blocks on each other. `"attachmentAllowed": false` will not allow that. But if I don't keep that, then the prefab that appears on activation will overlap with the other block placed on the root block. Please suggest what should I do.

File: assets/prefabs/testRook.prefab [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+    "skeletalmesh" : {
+            "mesh" : "rook",
```
[the remark this answers is not in the document]
Comment (darshan3) [reply assets/prefabs/testRook.prefab]: Yes, it comes from L&S Resources. It was just used for testing purposes as I didn't have any tower related prefab available. I will remove it once this module is ready.

File: src/main/java/org/terasology/logic/ingame/towers/SendTowerActivationRequest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+package org.terasology.logic.ingame.towers;
+
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.Event;
+import org.terasology.network.ServerEvent;
+
+@ServerEvent
+public class SendTowerActivationRequest implements Event {
```
[the remark this answers is not in the document]
Comment (darshan3) [reply src/main/java/org/terasology/logic/ingame/towers/SendTowerActivationRequest.java]: Yup, it is an activation request. I have renamed it as you have suggested.

File: src/main/java/org/terasology/logic/ingame/towers/TowerAuthoritySystem.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    private void deactivateTower(EntityRef towerEntity, TowerComponent towerComponent) {
+        towerComponent.childEntity.destroy();
+        towerComponent.childEntity = EntityRef.NULL;
+
+        if (delayManager.hasPeriodicAction(towerEntity, getActionId(towerEntity.getId()))) {
+            delayManager.cancelPeriodicAction(towerEntity, getActionId(towerEntity.getId()));
+        }
+    }
+
+    private String getActionId(long id) {
```
[the remark this answers is not in the document]
Comment (darshan3) [reply src/main/java/org/terasology/logic/ingame/towers/TowerAuthoritySystem.java]: Yup, I would add it once basic implementation is done. I was concerned about changes in the code and implementation post review, hence I have just added code for now. Once implementation is final, I will add javadocs everywhere as well as how to use this module in a README file.

File: src/main/java/org/terasology/logic/ingame/towers/TowerComponent.java (hunk 1 above)
Comment (darshan3) [reply src/main/java/org/terasology/logic/ingame/towers/TowerComponent.java]: Same as the previous comment.

**Review by iaronaraujo — APPROVED** [review]

**Pull request merged**